contracts/Convex.sol

```diff
@@ -0,0 +1,250 @@
+  uint256 public constant virtualPriceTol = scale - 500; // toggle if virtual price drops by >50%
+  uint256 public constant balanceTol = scale - 500; // toggle if true balances are >50% lower than internally tracked balances
+  address public constant convex = 0xF403C135812408BFbE8713b5A23a04b3D48AAE31; // Convex deposit contract (booster)
+
+  uint256 public immutable convexPoolId; // Convex deposit contract (booster) pool id
+  address public immutable convexToken; // Convex receipt token minted on deposits
+  address public immutable staker; // Convex contract that manages staking
+  address public immutable gauge; // Curve gauge that Convex deposits into
+
+  address public immutable curveMetaPool; // Curve meta pool
+  address public immutable curveBasePool; // Base Curve pool
+
+  address public immutable metaToken0;
+  address public immutable metaToken1;
+
+  address public immutable baseToken0;
+  address public immutable baseToken1;
+  address public immutable baseToken2;
+
+  uint256 public lastVpBasePool; // last virtual price read from base pool
+  uint256 public lastVpMetaPool; // last virtual price read from meta pool
+
+  function basePoolCoins(uint256 index) internal view virtual returns (address);
+
+  function basePoolBalances(uint256 index) internal view virtual returns (uint256);
+
+  /**
+   * @param _convexPoolId TODO
+   * @dev For definitions of other constructor parameters, see ITrigger.sol
+   */
+  constructor(
+    string memory _name,
+    string memory _symbol,
+    string memory _description,
+    uint256[] memory _platformIds,
+    address _recipient,
+    uint256 _convexPoolId
+  ) ITrigger(_name, _symbol, _description, _platformIds, _recipient) {
+    // Get addresses from the pool ID
+    (address _curveLpToken, address _convexToken, address _gauge, , , ) = IConvexBooster(convex).poolInfo(
+      _convexPoolId
+    );
+    staker = IConvexBooster(convex).staker();
+    convexPoolId = _convexPoolId;
+    convexToken = _convexToken;
+    gauge = _gauge;
+
+    curveMetaPool = ICrvToken(_curveLpToken).minter();
+    curveBasePool = ICrvMeta(curveMetaPool).base_pool();
+
+    metaToken0 = ICrvMeta(curveMetaPool).coins(0);
+    metaToken1 = ICrvMeta(curveMetaPool).coins(1);
+
+    baseToken0 = basePoolCoins(0);
+    baseToken1 = basePoolCoins(1);
+    baseToken2 = basePoolCoins(2);
+
+    // Get virtual prices
+    lastVpMetaPool = ICrvMeta(curveMetaPool).get_virtual_price();
+    lastVpBasePool = ICrvBase(curveBasePool).get_virtual_price();
+  }
+
+  function checkTriggerCondition() internal override returns (bool) {
+    // Typically in this method we check all conditions, save them to storage, amd return the result.
+    // This is convenient because it ensures we have the data that caused the trigger saved into
+    // the state, but this is just convenient and not a requirement. We do not follow that pattern
+    // here because  certain trigger conditions can cause this method to revert if we tried that
```

Comment:
nit: extra space between "because" and "certain"

Reply:
Fixed in fd73021